### ISSUE TITLE
Stream from Wiimote App getting wrong timestamps

### ISSUE THREAD

**Issue by WindyOrgUK** [labstreaminglayer/App-Wiimote#1]
Good evening,

XDF format is now being implemented in Brainstorm
(https://neuroimage.usc.edu/forums/t/new-version-with-xdf-file-support/36484/10)
But we just found out that the timestamps in WiimoteMarkers1 stream are being derived from 'created_at' timestamp:

sample from my test xdf file read by pyxdf python module:

The relevant part of "info"
            {'channel_count': ['1'],
             'channel_format': ['string'],
             'created_at': ['276643.5543640970'],
...
...
             'effective_srate': 0,
             'hostname': ['Sayonara.local'],
             'name': ['WiimoteMarkers1'],
             'nominal_srate': ['0.000000000000000'],
             'session_id': ['default'],
             'source_id': ['Wiimote_markers_1'],
             'stream_id': 2,
             'type': ['Markers'],
             'uid': ['add7be69-9fa8-4d04-a636-739d53197560'],
             'v4address': [None],
             'v4data_port': ['16572'],
             'v4service_port': ['16572'],
             'v6address': [None],
             'v6data_port': ['16572'],
             'v6service_port': ['16572'],
             'version': ['1.100000000000000']})

The content of "time_stamps":

array([277088.93317115, 277089.13319465, 277089.55193388, 277089.71946052,
       277090.07695277, 277090.24193617, 277090.63821009, 277090.74555763,
       277091.1505598 , 277091.31935779, 277091.62072755, 277091.74194948])

is this a bug (I hope) or a feature ?
This makes it impossible to relate the wiimote events to eeg data streamed into LabRecorder

Regards
Peter

**Comment by cboulay** [comment]
I just took a look at the source code and it looks correct to me. The timestamp being passed to push_sample is the return of lsl::local_clock, as it should be.
https://github.com/labstreaminglayer/App-Wiimote/blob/master/mainwindow.cpp#L384-L387

Each stream's timestamps should be derived from the same monotonic clock that LSL uses, most easily accessed with lsl::local_clock. (Note: providing NO timestamp will default to using this same clock).

Each computer has its own clock that runs independently of all other computers.

LabRecorder stores the data along with the originating timestamps. In addition, LabRecorder records the offsets between the originating computer's clock and the LabRecorder computer's clock.

When you load the xdf file with pyxdf or xdf-Matlab, a regression is used to fit the offsets and modify the originating timestamps so they are all in the same time base (i.e., that of the LabRecorder computer).

In other words, if you use pyxdf or xdf-Matlab with the default options then all streams should be synchronized after import.

If you find that your wiimote events and your EEG data have wildly different timestamps, then either you're telling pyxdf to not fix the clock offsets or there's a problem with the EEG implementation.

I'm going to close this issue because it's not an issue with the Wiimote app. I'll still respond here if you want to continue the conversation. I would need to know what the EEG app is before I can help you further.

**Comment by WindyOrgUK** [comment]
Thank you for your response.

The examples from my first post show exactly the values from the stream I captured yesterday.
The wiimote streams (WiimoteMarkers1 and wiiuse1) were the only 2 streams at this time.
The timestamps look just wrong (time from device power-up ?)
The python script just imports the raw data and just displays the values :

import pyxdf
from pprint import pprint

data, header = pyxdf.load_xdf('/Users/windy/wiimote_4.xdf')

pprint(data[0]["info"])
pprint(data[0]["time_stamps"])

Am I missing something obvious here ?

Peter

**Comment by cboulay** [comment]
> The timestamps look just wrong

Wrong according to what? What are you expecting?

The monotonic clock is the time elapsed since some arbitrary point in the past, maybe 1970-01-01, maybe since system restart. It's platform dependent.

> Am I missing something obvious here ?

After you've finished loading your streams, you can try something like the following (untested code).

```Python
t0 = min([min(_["time_stamps"]) for _ in streams])
for _ in streams:
    _["time_stamps"] -= t0
```

**Comment by WindyOrgUK** [comment]
So it seems that i made a wrong assumption that this is the labrecorder receiver/server application that timestamps the records and not the sending/client.
Is there a way to make labrecorder to timestamp the incoming records in this kind of situations where the client is not aware of the real time ?
I understand the risk of delays of course and am ready to accept it in this case.

**Comment by cboulay** [comment]
Why? What does that achieve? What are you trying to do here?

LabRecorder will give you the sender's timestamps AND the clock offsets between the sender and the LabRecorder computer. Then, when loading the xdf, the clock offsets will be compensated for. This is the best way to synchronize multiple streams from different computers!

Even if you only have 1 computer generating your streams, the approach that LabRecorder takes doesn't hurt anything.

Is it just that you want the timestamps in wall-clock time? The file header has a `datetime` field which will get you close: https://github.com/labstreaminglayer/App-LabRecorder/blob/master/xdfwriter/xdfwriter.cpp#L38 . But that just tells you when the file was opened, not when a stream was joined, so you can't assume that the first timestamp in a stream is close to the file-created datetime.
A better solution would be to write a custom app that creates a new outlet that pushes datetime representations at regular intervals, and save this stream along with the others in LabRecorder.

**Comment by WindyOrgUK** [comment]
Ok, let me simply read the docs as I do not want to waste your time and the solution may simply be somewhere there.
In brief, I need to correlate my button pushes on the wiimote with EEG recorded at the same time using for example OpenBCI and on top of that with events generated by psychopy.
3 different streams generated by 3 different devices/systems of which at least one (wiimote) does not have a clock that can be synched with anything else.

But again, let me RTFM first.

Regards
Peter

**Comment by cboulay** [comment]
Make sure that your LSL-psychopy implementation doesn't call `push_chunk` or `push_sample` with timestamps that are in a different time base; use pylsl.local_clock() -- possibly adjusted for a known frame delay -- or don't provide a timestamp at all and internally LSL will use the timestamp at the time of push.

Some older versions of the OpenBCI LSL plugin were sending incorrect timestamps but this has been fixed for over a year AFAIK.

Assuming you're OK on those 2 caveats, just try it out. Record all 3 streams into LabRecorder, then import the xdf file, and plot the data using the timestamps returned on xdf import. (The x-tick-labels will be gigantic so you can use the snippet I wrote above to reset them to a common zero.)

You can tap the wiimote against your EEG cap and see if the accelerometer data aligns with EEG artifacts.
